elfpy/markets/hyperdrive/hyperdrive_pricing_model.py

```diff
@@ -442,3 +448,282 @@ def calc_tokens_out_given_lp_in(
+            #
+            # We can approximate the change in the trade output with respect to
+            # trade size as dy' = c * (1/p) * dz'. Substituting this into our error
+            # equation and setting the error equation equal to zero allows us to
+            # solve for the trade size update:
+            #
+            # (y_l + dy + c * (1/p) * dz') / c - (z + dz + dz') = 0
+            #                  =>
+            # (1/p - 1) * dz' = (z + dz) - (y_l + dy) / c
+            #                  =>
+            # dz' = ((z + dz) - (y_l + dy) / c) * (p / (p - 1)).
+            p = calculate_spot_price(share_reserves + dz, bond_reserves - dy, initial_share_price, time_stretch)
+            error = int((share_reserves + dz) - (longs_outstanding + dy) / share_price)
+            if error > 0 and dz * share_price > result.base_amount:
+                result = MaxLongResult(base_amount=dz * share_price, bond_amount=dy)
+            if p >= ONE_18:
+                break
+            if error < 0:
+                dz -= -error * p / (ONE_18 - p)
+            else:
+                dz += error * p / (ONE_18 - p)
+            dy = calculate_bonds_out_given_shares_in(
+                share_reserves, bond_reserves, dz, ONE_18 - time_stretch, share_price, initial_share_price
+            )
+
+        return result
+
+    def calculate_max_short(
+        self,
+        share_reserves: FixedPoint,
+        bond_reserves: FixedPoint,
+        longs_outstanding: FixedPoint,
+        time_stretch: FixedPoint,
+        share_price: FixedPoint,
+        initial_share_price: FixedPoint,
+        minimum_share_reserves: FixedPoint,
+    ) -> FixedPoint:
+        r"""
+        Calculates the maximum amount of shares that can be used to open shorts.
+
+        Parameters
+        ----------
+        share_reserves : FixedPoint
+            The pool's share reserves.
+        bond_reserves : FixedPoint
+            The pool's bonds reserves.
+        longs_outstanding : FixedPoint
+            The amount of longs outstanding.
+        time_stretch : FixedPoint
+            The time stretch parameter.
+        share_price : FixedPoint
+            The share price.
+        initial_share_price : FixedPoint
+            The initial share price.
+
+        Returns
+        -------
+        FixedPoint
+            The maximum amount of shares that can be used to open shorts.
+        """
+        t = ONE_18 - time_stretch
+        price_factor = share_price / initial_share_price
+        k = modified_yield_space_constant(price_factor, initial_share_price, share_reserves, t, bond_reserves)
+        inner_factor = initial_share_price * longs_outstanding / share_price + minimum_share_reserves**t
+        optimal_bond_reserves = (k - price_factor * inner_factor) ** (ONE_18 / t)
+
+        # The optimal bond reserves imply a maximum short of dy = y - y0.
+        return optimal_bond_reserves - bond_reserves
+
+
+class MaxLongResult(NamedTuple):
```

Comment:
I think you should move this and `ONE_18` to the top of this file or to their own file